dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/OpenTelemetryInfoFactory.java

```diff
@@ -0,0 +1,357 @@
+     *    If invoked after an app has been shutdown these will return a no-op object.
+     * 
+     * 2. A cache of OpenTelemetryInfo (using J2EENames as the key). OpenTelemetryInfo in turn contains OpenTelemetry objects. This is populated lazily using computeIfAbsent and the use of ConcurrentHashMap ensures thread safety.
+     *    
+     * 3. An ApplicationStateListener. 
+     *    On startup the ASL only removes applications from the list of apps that have shut down so they will not return a no-op
+     *    On shutdown the ASL cleans up an OpenTelemetry object, removes its parent OpenTelemetryInfo from the cache, and registers it as an app which has shut down.
+     */
+
+    private static final String INSTRUMENTATION_NAME = "io.openliberty.microprofile.telemetry";
+    private static final String ENV_DISABLE_PROPERTY = "OTEL_SDK_DISABLED";
+    private static final String CONFIG_DISABLE_PROPERTY = "otel.sdk.disabled";
+    private static final String ENV_METRICS_EXPORTER_PROPERTY = "OTEL_METRICS_EXPORTER";
+    private static final String CONFIG_METRICS_EXPORTER_PROPERTY = "otel.metrics.exporter";
+    private static final String ENV_LOGS_EXPORTER_PROPERTY = "OTEL_LOGS_EXPORTER";
+    private static final String CONFIG_LOGS_EXPORTER_PROPERTY = "otel.logs.exporter";
+    private static final String SERVICE_NAME_PROPERTY = "otel.service.name";
+
+    private static final TraceComponent tc = Tr.register(OpenTelemetryInfoFactory.class);
+
+    private static final Map<J2EEName, OpenTelemetryInfo> appToInfo = new ConcurrentHashMap<J2EEName, OpenTelemetryInfo>();
+    private static final Set<J2EEName> shutDownApplications = Collections.newSetFromMap(new WeakHashMap<J2EEName, Boolean>()); //We have no good place to say forget about an app entirely, so I use weakreferences. 
+
+    //The key methods that provide access to telemetry objects begin here
+    public static OpenTelemetryInfo getOpenTelemetryInfo(J2EEName j2EEName) {
+        return appToInfo.computeIfAbsent(j2EEName, OpenTelemetryInfoFactory::createOpenTelemetryInfo);
+    }
+
+    public static Tracer getTracer(J2EEName j2EEName) {
+        try {
+            return getOpenTelemetryInfo(j2EEName).getOpenTelemetry().getTracer(INSTRUMENTATION_NAME);
+        } catch (Exception e) {
+            Tr.error(tc, Tr.formatMessage(tc, "CWMOT5002.telemetry.error", e));
+            return OpenTelemetry.noop().getTracerProvider().get("");
+        }
+    }
+
+    //The key methods that provide access to telemetry objects end here
+
+    private static OpenTelemetryInfo createOpenTelemetryInfo(J2EEName j2EEName) {
+        //If the app has already shut down but something still tries to get OpenTelemetry objects return a no-op object
+        if (shutDownApplications.contains(j2EEName)){ //There is a low risk race condition here. Thread A gets past this check just as thread B shuts everything down.
```

Comment:
As a followup: fix this race condition.

We shouldn't allow a new `OpenTelemetryInfo` to be created after the app has been shut down because at that point we won't clean it up again.

Reply:
I did think of a way to make it clearer that this code is safe:

- Add a new class `OpenTelemetryInfoHolder`, with one field which is the `OpenTelemetryInfo` and another which is `boolean disposed`
- Change `appToInfo` to type `Map<J2EEName, OpenTelemetryInfoHolder>`
- On app start, add an empty holder to the map
- On `getOpenTelemetryInfo`, look up the holder from the map.
  - Sync on the holder.
  - If the holder is empty, check it hasn't been disposed, then create an `OpenTelemetryInfo` and store it in the holder
- On app shutdown, remove the holder from the map
  - Sync on the holder
  - If the removed holder is not empty, dispose of the `OpenTelemetryInfo`
  - Mark the holder disposed

This avoids having a separate map to track shut down apps and provides an easy object to synchronize on at the cost of an extra level of indirection when doing the lookup.

Reply:
Tom also had the suggestion that we could store the `OpenTelemetryInfoHolder` in the app metadata itself, rather than using the app name as a key to a map.

Reply:
Lets discuss this one.